### ISSUE TITLE
Enable engine to support to mirror private registry

### ISSUE THREAD

**Issue by TrumanLing** [docker/docker#18818]
Now I'm using DTR(docker trusted registry) on trial, and setup mirror for DTR,but engine can't support to mirror private registry. So, I think engine need to enable this function, how about your opinion? I have implemented it, if you agree with that, I will make a PR.

**Comment by thaJeztah** [comment]
Looks like you also opened this issue here; https://github.com/docker/docker/issues/18818

**Comment by thaJeztah** [comment]
Actually, I meant to link to https://github.com/docker/docker/issues/16974, which should be on the registry roadmap; https://github.com/docker/distribution/pull/1136

**Comment by TrumanLing** [comment]
@thaJeztah I understand your comment

But as dmp42's description in docker/distribution#922, distribution itself supports private registry, but engine not yet.So, here I submit another issue for engine to support this feature.

**Comment by saiwl** [comment]
I think it's very useful!

**Comment by TrumanLing** [comment]
The corresponding PR is #18915

**Comment by kinglion811** [comment]
Is this PR fit in with this https://github.com/docker/docker-registry/issues/1075

**Comment by llamahunter** [comment]
What's the progress on this issue?  It seems strange that pull through proxy mirrors are supported for *one* registry (docker.io), but not any other registries.

**Comment by thaJeztah** [comment]
ping @stevvooe @dmp42 ptal

**Comment by dubo-dubon-duponey** [comment]
Mirroring private repositories is desirable, but is not as simple as "just enable it".

How would image name resolution work? Will `docker pull ubuntu` change meaning depending on what you are caching and what not? How does that work with authentication?

The issue tracking this is here: https://github.com/docker/distribution/issues/1431

**Comment by dmp42** [comment]
Correct.

@llamahunter 

> What's the progress on this issue? It seems strange that pull through proxy mirrors are supported for one registry (docker.io), but not any other registries.

The reason for that is indeed about namespace consistency.

If you have private registry on domain A mirror private registry on domain B, then:

`docker pull domainB/foo` could really mean something different for different location / configuration, while it's not the case for short name images (speaking under @RichardScothern control).

A fragmentation of the image namespace would be extremely bad for the community, and break one of docker fundamental assumption (build producing the same result).

**Comment by dmp42** [comment]
@llamahunter yeah, I understand the need.

Couple of thoughts:

 * images are fundamentally http cacheable objects, so, my expectation is that using a plain HTTP caching proxy (and configuring your system to use it) would just work (this has to be tested though).
 * the registry supports a delivery delegation mechanism, that let you use a CDN (cloudfront, fastly) to handle the layers delivery part instead of your "central" registry - that might be an option too 

For the rest, I encourage you to channel the discussion into the issue that was linked earlier ( https://github.com/docker/distribution/issues/1431 ) since this is where most of the discussion is happening.

Hope that helps.

**Comment by llamahunter** [comment]
I've commented on the 'master issue' 4 months ago, but not much seems to be happening there.

**Comment by virtuald** [comment]
@dmp42 you're right, you can use squid to do this, I haven't tested it extensively but it seems to work at first blush. 

@llamahunter see https://github.com/virtuald/docker-registry-cache for a HTTP cache using squid

**Comment by jmahowald** [comment]
I was considering using GeoDNS to handle this (possibly with that squid cache from @virtuald .  Wondering if anyone on this thread knows reasons that might not work.

**Comment by vrothberg** [comment]
I'd like to revive this discussion with a concrete pull request (see https://github.com/moby/moby/pull/34319). If you find time, please take a look at our proposal and share your thoughts and concerns. In the following, I try to give a small summary of the PR, but please refer to the PR and commit message (see https://github.com/moby/moby/pull/34319) for detailed information:

The ultimate goal of the PR is to improve the user experience by allowing faster image pulls from private registries. A private-registry mirror can be added either via the daemon cli or by adding an entry to `/etc/docker/daemon.json` such as:
```json
{
  "registry-mirrors": ["https://hub_mirror.local.lan"],
  "private-registry-mirrors": {"containers.scc.suse.com": "https://scc_mirror.local.lan"}
}
```
With this PR and the above configuration the following command would work as expected:
```
$ docker pull opensuse
$ docker pull containers.scc.suse.com/sles12sp3
$ docker pull containers.scc.suse.com/caasp/mariadb
```
Notice that we make sure that a mirror can serve only one registry.

Ping: @thaJeztah, @dmp42, @stevvooe

**Comment by duyanghao** [comment]
@thaJeztah any progress for docker supporting private registry mirror?
It is a very important feature but seems to have received only scant attention and move ahead slowly : )

**Comment by vrothberg** [comment]
@duyanghao We are working on https://github.com/moby/moby/pull/34319 and converge to a concrete design and implementation. I hope we can get it merged asap.

**Comment by duyanghao** [comment]
@vrothberg Nice work, i will have a look at this PR.

**Comment by vrothberg** [comment]
> @vrothberg Nice work, i will have a look at this PR.

I am currently updating the PR. I'll let you know in #34319 once that's done. Just to avoid you wasting time reading code that is just about to change.

**Comment by qct** [comment]
guys, any updates about this issue ?  looks @vrothberg did a perfect work

**Comment by vrothberg** [comment]
> guys, any updates about this issue ? looks @vrothberg did a perfect work

Thanks for the kind words! As I changed employer, I am not pushing the PR forward anymore. But there is no need to worry as I coordinated that with the maintainers and my former colleagues (@flavio, @cyphar and @marcov) who will take over and hopefully guide the feature over the finish line.

**Comment by flavio** [comment]
Yes, we plan to resume the work on this PR after the holiday seasons. We fear that resuming the discussion right now would not be productive due to a lot of people being on vacation.

**Comment by mrgavinconway001** [comment]
Any update on this?

**Comment by qct** [comment]
I think holiday seasons should have passed.  LOL  :smile:

**Comment by aven-ai** [comment]
Any update on this?  Six years passed, is there any plan to support private registry?

**Comment by cruzanstx** [comment]
Any update?

**Comment by shulinwz** [comment]
Just switch to Podman. Setting up a private registry cache is dead simple with Harbor and containers-registries.conf.

**Comment by ydcool** [comment]
ping

**Comment by timorkal** [comment]
somebody?

**Comment by dali546** [comment]
Hi, I am also interested in this feature. Having mirror support for registries beyond DockerHub vastly can improve latency when using other registries like gcr, ghcr, ecr etc

**Comment by navarro967** [comment]
There are way too many issues, requests, comments, use cases, removal from milestones, and closed PRs for this feature. I understand though, I myself do not currently have the time needed to dedicate to resolving this. On paper it seems insane that something that sounds simple enough hasn't been done in nearly 9 years.

If you, reading this, are a aka bored developer who enjoys contributing to open-source projects and has free time, we need you.

**Comment by jwillikers** [comment]
> Just switch to Podman. Setting up a private registry cache is dead simple with Harbor and containers-registries.conf.

Agreed. If you don't need support for Windows containers, Podman's support for using mirrors for any registry is really slick and pretty easy to configure.

**Comment by Vigilans** [comment]
I found that with [containerd snapshotter](https://docs.docker.com/engine/storage/containerd/) mode enabled, docker engine will be able to configure mirror for any registry using [containerd's hosts configuration](https://github.com/containerd/containerd/blob/main/docs/hosts.md) in `/etc/docker/certs.d` directory.

Example 1: `hosts.toml` for quay.io (`/etc/docker/certs.d/quay.io/hosts.toml`):
```toml
[host."https://quay.m.daocloud.io"]
  capabilities = ["pull", "resolve"]
```

Example 2: `hosts.toml` for harbor's pull through cache:
```toml
[host."http://harbor.domain/v2/quay.io"]
  capabilities = ["pull", "resolve"]
  override_path = true
```

Example 3: `hosts.toml` using nexus registry for pulling and pushing (assume `quay.io` is a group repository for `quay.io-hosted` and `quay.io-proxy`:
```toml
server = "http://127.0.0.1:8081/repository/quay.io"

[host."http://127.0.0.1:8081/repository/quay.io"]
  capabilities = ["pull", "resolve"]
  header.Authorization = "Basic <base64 encoded user:password>"

# `http://127.0.0.1:8081/repository/quay.io-hosted` will not work properly as nexus stripped /repository/<registry> from the path in returned blob url
[host."http://127.0.0.1:8082"] # Http connector port of quay.io-hosted
  capabilities = ["push"]
  header.Authorization = "Basic <base64 encoded user:password>"
```

Notes:
1. This feature is enabled in a recent version of docker. I found it is not available in docker-ce 26.1.3, while available since 28.0.0. I haven't check what is the exact commit that introduced it.
2. The hosts directory for docker engine is `/etc/docker/certs.d`. In some other tutorials you may find another directory `/etc/containerd/certs.d`, this is for containerd **cri plugin** only. Containerd itself (using `ctr` command) does not use any configuration file, instead it requires you to explicitly provide hosts dir as argument (using `--hosts-dir`).

However, after trial for months, I found there are caveats that cannot be ignored when enabling containerd snapshotter(since it is still experimental now):
1. Storage driver options, like setting quota for container, is not available. See:
   * https://github.com/containerd/containerd/issues/10475
   * https://github.com/containerd/containerd/issues/759.
2. When pulling images, the pulled image will occasionally miss some layers ("occasionally" because I didn't find a way to stably reproduce it, but I think there is a way). The `created` column in `docker image ls` shows `55 years ago` or so. Pull using `ctr -n moby image pull` directly will fix the image. The cause seems to be related to following issues:
   * https://github.com/moby/moby/issues/49784
   * https://github.com/moby/moby/issues/49473

**Comment by Sam13** [comment]
@Vigilans Nice, thank you. Tried as well, works great.
Did you find nicer way to configure the credentials for a host?
According my experiments `docker login` for that specific host does not work.

**Comment by Vigilans** [comment]
@Sam13 See the `header.Authorization = "Basic <base64 encoded user:password>"` line.

**Comment by Sam13** [comment]
> `header.Authorization = "Basic <base64 encoded user:password>"`

@Vigilans Thanks, I'm aware of that and it works.
A smarter way without hardcoded and unencrypted credentials would be nice...

**Comment by Vigilans** [comment]
> A smarter way without hardcoded and unencrypted credentials would be nice...

Ummm, there is a non-hardcoded but stupid way...

You can docker login to the mirrored host first, then manually edit `~/.docker/config.json` to rename/copy-paste the section of your mirror to your target registry (for docker it seems to be `https://index.docker.io/v1/`, cannot remember what I was exactly using).

**Comment by Sam13** [comment]
If I have the correct credentials for my mirror in `~/.docker/config.json` (see example) I still get the following message from Docker daemon  
```
msg="trying next host" error="pull access denied, repository does not exist or may require authorization: authorization failed: no basic auth credentials" host=mymirror.domain.ext
```

Example:
_hosts.toml_
```toml
[host."https://mymirror.domain.ext"]
  capabilities=["pull", "resolve"]
```
_~/.docker/config.json_
```json
{
        "auths": {
                "mymirror.domain.ext": {
                        "auth": <encoded token>
                }
        }
}
```

**Comment by Vigilans** [comment]
> {
>         "auths": {
>                 "mymirror.domain.ext": {
>                         "auth": <encoded token>
>                 }
>         }
> }

You need to ensure auth entry for your **actual** registry the `docker` command is interacting with exists in `~/.docker/config.json`. E.g. If you are pulling docker hub images (proxied through `mymirror.domain.ext`), you need to ensure auth entry for `https://index.docker.io/v1/` exists, instead of `mymirror.domain.ext`, because docker cli will attatch Authorization header only based on `docker.io`, as a cli it does not know `docker.io` is proxied by another mirror in daemon.

**Comment by Sam13** [comment]
> You need to ensure auth entry for your **actual** registry the `docker` command is interacting with exists in `~/.docker/config.json`. 

That doesn't work in my case: No matter what I have authenticated in my `~/.docker/config.json` the only thing which does work is the authentication header configuration in `hosts.toml`.

Without header configuration in `hosts.toml`:
If I'm logged in to the actual registry the mirror is not respected.
If I'm logged in to the mirror registry the credentials are not used for the mirror.

My use case is to mirror a private registry from the cloud to a local one where the same authentication can be used for both registries.
